### PR TITLE
gh-93162: Add ability to configure QueueHandler/QueueListener together and provide getHandlerByName() and getHandlerNames() APIs.

### PR DESCRIPTION
**Pull request by vsajip**
…ide a getHandlerByName API.

### CODE REVIEW

File: Doc/library/logging.config.rst [hunk 1]
```diff
@@ -661,6 +661,61 @@ it with :func:`staticmethod`. For example::
...
+
+Configuring QueueHandler and QueueListener
+""""""""""""""""""""""""""""""""""""""""""
+
+If you want to configure a :class:`~logging.handlers.QueueHandler`, noting that this
+is normally used in conjunction with a :class:`~logging.handlers.QueueListener`, you
+can configure both together. After the configuration, the ``QueueListener`` instance
+will be available as the :attr:`listener` attribute of the created handler, and that
+in turn will be available to you using :func:`~logging.getHandlerByName` and passing
+whatever name you have used for the ``QueueHandler`` in your configuration. The
```
Comment (iritkatriel) [inline Doc/library/logging.config.rst]:
```suggestion
the name you have used for the ``QueueHandler`` in your configuration. The
```

File: Doc/library/logging.config.rst [hunk 2]
```diff
@@ -661,6 +661,61 @@ it with :func:`staticmethod`. For example::
...
+  code.
+
+* A string that resolves to a callable which, when called with no arguments, returns
+  the :class:`queue.Queue` instance to use. That callable could be a
+  :class:`queue.Queue` subclass or a function which returns a suitable queue instance.
+
+If the  ``queue`` key is absent, a standard unbounded :class:`queue.Queue` instance is
+created and used.
+
+If the ``listener`` key is present, the corresponding value can be one of the following:
```
Comment (iritkatriel) [inline Doc/library/logging.config.rst:710]: You say "one of the following" and then there is only one listed?

File: Doc/library/logging.config.rst [hunk 3]
```diff
@@ -661,6 +661,61 @@ it with :func:`staticmethod`. For example::
...
+  possible if you are constructing or modifying the configuration dictionary in
+  code.
+
+If the ``listener`` key is absent, :class:`logging.handlers.QueueListener` is used.
+
+The values under the ``handlers`` key are the names of other handlers in the
+configuration (not shown in the above snippet) which will be passed to the queue
+listener.
+
+Any custom queue handler and listener classes will need to deal with the same
```
Comment (iritkatriel) [inline Doc/library/logging.config.rst]:
```suggestion
Any custom queue handler and listener classes will need be defined with the same
```

File: Lib/logging/config.py [hunk 4]
```diff
@@ -721,26 +738,61 @@ def configure_handler(self, config):
...
+                        config['listener'] = listener
+                    elif not issubclass(lspec, logging.handlers.QueueListener):
+                        raise TypeError('Invalid listener spec %r' % lspec)
+                hlist = []
+                try:
+                    for hn in config['handlers']:
+                        h = self.config['handlers'][hn]
+                        if not isinstance(h, logging.Handler):
+                            config.update(config_copy)  # restore for deferred cfg
+                            raise TypeError('needed handler %r '
```
Comment (iritkatriel) [inline Lib/logging/config.py]:
```suggestion
                            raise TypeError('Required handler %r '
```

File: Lib/logging/config.py [hunk 5]
```diff
@@ -563,7 +565,7 @@ def configure(self):
                         handler.name = name
                         handlers[name] = handler
                     except Exception as e:
-                        if 'target not configured yet' in str(e.__cause__):
+                        if ' not configured yet' in str(e.__cause__):
```
Comment (iritkatriel) [inline Lib/logging/config.py:568]: It would be better to define an exception type for this rather than use the text of the exception message for control flow (this should be done in a separate PR as it's not a new issue here).

**Review by iritkatriel — APPROVED** [review]

**Comment by vsajip** [comment]
@iritkatriel Thanks for the comments, I agree that these are all helpful suggestions / improvements. I will make the changes soon.

**Comment by vsajip** [comment]
This should fix #93162

File: Doc/library/logging.config.rst [hunk 6]
```diff
@@ -661,6 +661,76 @@ it with :func:`staticmethod`. For example::
...
+* A subclass of :class:`logging.handlers.QueueListener`. This is of course only
+  possible if you are constructing or modifying the configuration dictionary in
+  code.
+
+* A string which resolves to a class which is a subclass of ``QueueListener``, such as
+  ``'my.package.CustomListener'``.
+
+* A dict with a ``'()'`` key which is constructed in the usual way as discussed in
+  :ref:`logging-config-dict-userdef`. The result of this construction should be a
+  callable with the same signature as the ``QueueListener`` initializer.
```
Comment (spacemanspiff2007) [inline Doc/library/logging.config.rst:721]: ``The result of this construction should be a callable with the same signature...``

Does that mean ``()`` will be resolved, then called with the dict args and the result of that call should return a callable again?

Comment (spacemanspiff2007) [inline Doc/library/logging.config.rst:721]: ``The result of this construction should be a callable with the same signature...``

Does that mean ``()`` will be resolved, then called with the dict args and the result of that call should return a callable again?

Comment (vsajip) [reply Doc/library/logging.config.rst:721]: Yes. The callable will be called just as `QueueListener` would be.

**Comment by vsajip** [comment]
I just deleted a duplicated comment.

File: Doc/library/logging.config.rst (hunk 6 above)
Comment (spacemanspiff2007) [reply Doc/library/logging.config.rst:721]: But that's different than the normal behavior of '()':
````
In order to provide complete flexibility for user-defined object instantiation, the user needs to provide a ‘factory’ - 
a callable which is called with a configuration dictionary and which returns the instantiated object.
````
[from docs](https://docs.python.org/3/library/logging.config.html?highlight=dictconfig#user-defined-objects).
Note that the return normally is not a callable but an instantiated object.

Comment (vsajip) [reply Doc/library/logging.config.rst:721]: Yes, but in this case the value to be returned is documented explicitly. How can it be returned as an instantiated object in this case, if it has no access to the queue?

Comment (vsajip) [reply Doc/library/logging.config.rst:721]: Also note that "instantiated object" is a generic term. It does not imply any specific type of object, and a callable is also an instantiated object.

Comment (spacemanspiff2007) [reply Doc/library/logging.config.rst:721]: Do you think it would be possible to provide an example in the docs?
That'll make things more clear.

Comment (vsajip) [reply Doc/library/logging.config.rst:721]: Feel free to submit a documentation PR with what you think should be in there, and I'll take a look.
